### PR TITLE
Add API Documentation in core package

### PR DESCRIPTION
**Pull request by sdirix**
* Document 'BackendContribution'
* Document 'MessageService' and related message interfaces
* Extend documentation of MenuContribution and related menu interfaces
and classes
* Document 'PreferenceService', 'PreferenceProxy' and related
interfaces, classes and functions

Signed-off-by: Stefan Dirix <sdirix@eclipsesource.com>
Contributed on behalf of STMicroelectronics

### CODE REVIEW

**Comment by sdirix** [comment]
I created only one commit as all of these files are in the core package.

`preference-proxy.ts`, `preference-service.ts`, `backend-application` -> Jonas
`menu.ts` -> Tobias
`message-service.ts`, `message-service-protocol.ts` -> Philip

@JonasHelming Note in `preference-service.ts` I did not only add documentation but introduced a new type (for documentation purposes) as well as changed one return type which was manually typed. I don't think this should pose any problems but is of course more risky than only adding documentation.

File: packages/core/src/common/menu.ts [hunk 1]
```diff
@@ -82,16 +125,37 @@ export class MenuModelRegistry {
         }
     }
```
Comment (tortmayr) [inline packages/core/src/common/menu.ts:127]: Maybe some short summary describing the purpose of this class?

**Review by tortmayr — COMMENTED** [review]
[Menu] In general great documentation! I added a few inline suggestions/comments

File: packages/core/src/browser/preferences/preference-proxy.ts [hunk 2]
```diff
@@ -12,6 +12,10 @@
  * https://www.gnu.org/software/classpath/license.html.
  *
  * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ * 
+ * Contributors:
+ *     TypeFox and others - Initial API and implementation
+ *     Stefan Dirix <sdirix@eclipsesource.com> (Contributed on behalf of STMicroelectronics) - Documentation
```
Comment (JonasHelming) [inline packages/core/src/browser/preferences/preference-proxy.ts:18]: omit "()"

File: packages/core/src/browser/preferences/preference-service.ts [hunk 3]
```diff
@@ -12,6 +12,10 @@
  * https://www.gnu.org/software/classpath/license.html.
  *
  * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ * 
+ * Contributors:
+ *     TypeFox and others - Initial API and implementation
+ *     Stefan Dirix <sdirix@eclipsesource.com> (Contributed on behalf of STMicroelectronics) - Documentation
```
Comment (JonasHelming) [inline packages/core/src/browser/preferences/preference-service.ts:18]: omit "()"

File: packages/core/src/common/menu.ts [hunk 4]
```diff
@@ -12,6 +12,10 @@
  * https://www.gnu.org/software/classpath/license.html.
  *
  * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ * 
+ * Contributors:
+ *     TypeFox and others - Initial API and implementation
+ *     Stefan Dirix <sdirix@eclipsesource.com> (Contributed on behalf of STMicroelectronics) - Documentation
```
Comment (JonasHelming) [inline packages/core/src/common/menu.ts:18]: omit "()"

File: packages/core/src/node/backend-application.ts [hunk 5]
```diff
@@ -12,6 +12,10 @@
  * https://www.gnu.org/software/classpath/license.html.
  *
  * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ * 
+ * Contributors:
+ *     TypeFox and others - Initial API and implementation
+ *     Stefan Dirix <sdirix@eclipsesource.com> (Contributed on behalf of STMicroelectronics) - Documentation
```
Comment (JonasHelming) [inline packages/core/src/node/backend-application.ts:18]: omit "()"

**Review by JonasHelming — APPROVED** [review]

File: packages/core/src/browser/preferences/preference-proxy.ts [hunk 6]
```diff
@@ -33,21 +37,99 @@ export interface PreferenceEventEmitter<T> {
...
+ * 
+ */
 export type PreferenceProxy<T> = Readonly<T> & Disposable & PreferenceEventEmitter<T> & PreferenceRetrieval<T>;
 export interface PreferenceProxyOptions {
+    /**
+     * Prefix which is transparently added to all preference identifiers.
+     */
     prefix?: string;
+    /**
+     * The default resourceUri to use when none was specified when calling "set" or "get".
```
Comment (tortmayr) [inline packages/core/src/browser/preferences/preference-proxy.ts:105]: `to use if none was specified when calling`

File: packages/core/src/common/menu.ts [hunk 7]
```diff
@@ -20,15 +24,31 @@ import { CommandRegistry, Command } from './command';
 import { ContributionProvider } from './contribution-provider';
 
 export interface MenuAction {
```
[the remark this answers is not in the document]
Comment (sdirix) [reply packages/core/src/common/menu.ts:29]: added summary

File: packages/core/src/common/menu.ts [hunk 8]
```diff
@@ -20,15 +24,31 @@ import { CommandRegistry, Command } from './command';
...
      * In addition to the mandatory command property, an alternative command can be defined.
      * It will be shown and invoked when pressing Alt while opening a menu.
      */
     alt?: string;
+    /**
+     * A specific label for this action. If not specified the command label or command id will be used.
+     */
     label?: string
+    /**
+     * Icon class(es). If not specified the command icon class will be used if it exists.
```
[the remark this answers is not in the document]
Comment (sdirix) [reply packages/core/src/common/menu.ts]: clarified meaning

File: packages/core/src/common/menu.ts [hunk 9]
```diff
@@ -191,6 +252,9 @@ export class MenuModelRegistry {
         return newSub;
     }
 
+    /**
+     * Returns the root menu or the menu specified by the given path.
```
[the remark this answers is not in the document]
Comment (sdirix) [reply packages/core/src/common/menu.ts]: good point! expanded description

File: packages/core/src/common/menu.ts (hunk 1 above)
Comment (sdirix) [reply packages/core/src/common/menu.ts:137]: done

**Review by tortmayr — APPROVED** [review]
[Menu] LGTM. :+1:

File: packages/core/src/browser/preferences/preference-service.ts [hunk 10]
```diff
@@ -66,34 +70,143 @@ export interface PreferenceChanges {
 }
 
 export const PreferenceService = Symbol('PreferenceService');
+/**
+ * Service to store and retrieve preference values.
+ * 
+ * Depending on your use case you might also want to look at {@link createPreferenceProxy} with which
```
Comment (tortmayr) [inline packages/core/src/browser/preferences/preference-service.ts:76]: `@link createPreferenceProxy`
Just ouf of curiosity: Does this work? Can TSDoc resolve crossreferences to another File?

Comment (sdirix) [reply packages/core/src/browser/preferences/preference-service.ts]: Good question! Probably not

**Comment by sdirix** [comment]
Reverted the copyright header changes

**Review by planger — APPROVED** [review]

**Comment by sdirix** [comment]
Split into #11, #12, #13 and #14

**Pull request closed without merging**